src/App.js

```diff
@@ -1,13 +1,126 @@
+    };
+  }
+
+  handleNewToDo = e => {
+    e.preventDefault();
+    // !this.state.input is equivalent to this.state.input === '' (falsey)
+    if (!this.state.input) {
+      this.setState({ invalidInput: true });
+    } else {
+      const task = {
+        // slice creates a copy of this.state.input
+        task: this.state.input.slice(),
+        id: Date.now(),
+        completed: false
+      };
+
+      // copy todos and add task to todos
+      const todos = [...this.state.todos, task];
+
+      // add task to todo list
+      // todos.push(task);
+
+      this.setState({ todos: todos, input: '' });
+
+      // Local Storage
+      localStorage.setItem('todos', JSON.stringify(todos));
+      localStorage.setItem('input', '');
+    }
+  };
+
+  handleInputChange = e => {
+    // if warning message is showing - hide it!
+    // eslint-disable-next-line
+    this.state.invalidInput ? this.setState({ invalidInput: false }) : null;
+    this.setState({ input: e.target.value });
+    localStorage.setItem('input', e.target.value);
+  };
+
+  // map over existing array with conditional
+  toggleCompleteTask = e => {
+    const newTodos = this.state.todos.map(todo => {
+      if (String(todo.id) === e.target.id) {
+        return {
+          ...todo,
+          completed: !todo.completed
+        };
+      } else {
+        return todo;
+      }
+    });
+    this.setState({ todos: newTodos });
+    localStorage.setItem('todos', JSON.stringify(newTodos));
+  };
+
+  handleCompletedTasks = e => {
+    e.preventDefault();
+    const removeCompleted = this.state.todos.filter(todo => !todo.completed);
+    this.setState({ todos: removeCompleted });
+
+    // Local Storage
+    localStorage.setItem('todos', JSON.stringify(removeCompleted));
+  };
+
+  // retrive todos from local storage
+  hydrateStateWithLocalStorage() {
+    // for all items in state
+    for (let key in this.state) {
+      // if the key exists in localStorage
+      if (localStorage.hasOwnProperty(key)) {
+        // get the key's value from localStorage
+        let value = localStorage.getItem(key);
+        // parse the localStorage string and setState
+        try {
+          value = JSON.parse(value);
+          this.setState({ [key]: value });
+        } catch (e) {
+          // handle empty string
+          this.setState({ [key]: value });
+        }
+      }
+    }
+  }
+
+  // React.Component method
+  componentDidMount() {
+    this.hydrateStateWithLocalStorage();
+  }
+
   render() {
     return (
-      <div>
-        <h2>Welcome to your Todo App!</h2>
+      <div className="App">
+        <TodoHeader />
+        <TodoForm
+          handleNewToDo={this.handleNewToDo}
+          handleInputChange={this.handleInputChange}
+          handleCompletedTasks={this.handleCompletedTasks}
+          input={this.state.input}
+        />
+        <TodoMessage
+          inlineStyle={this.state.invalidInput ? this.state.showMessage : null}
```

Comment:
Love it!

Reply:
However, the specific pattern of having this.state.showMessage on state, not sure the reason? It never gets updated, and I'm not sure if you would want to update it. What I would do is just have a CSS class called `.hidden`  that sets it to `display: none` and have a prop on TodoMessage that checks for whether invalidInput is on state, and use that prop to determine whether to include 'hidden' as className inside of TodoMessage by using a short circuit operator. Like this:

`<TodoMessage visible={Boolean(this.state.invalidInput)} />`

and then inside do:

```
const TodoMessage = props => {
  return (
    <div className="TodoMessage" className={props.visible && 'hidden'} >
      <i className="fas fa-exclamation-triangle" />
      Enter a task, s'il vous plaît!
    </div>
  );
};
```